#134: add Meteor Knight Fight

doubledh opened the pull request:
Summary
-------
Not quite sure how to go about balancing this, also the Knight has the right Predict tools against the Meteor Knight but... Meteor Knight has the type effectiveness 😅 

Was running into an exception during Knight predicts... I'm not entirely sure it's from my code (I think it has to do something with stuns)

Local Tests Performed
---------------------
- [ x ] bot still turns on (no BuildErrors or circular dependencies)
- [ ] (how did you ensure these changes don't break the code base and work as intended?)

Issue
-----
Closes #134

Commit:
issue134: add meteorknight fight

Commit:
issue 134: add meteorknight fight

ntseng commented on source/enemies/asteroid.js:
```diff
@@ -0,0 +1,73 @@
+      damage *= 2;
+    }
+    changeStagger(targets, "elementMatchFoe");
+    return `${dealDamage(targets, user, damage, false, user.element, adventure) + dealDamage([user], user, recoilDmg, false, "Untyped", adventure)}`;
```
```suggestion
    return `${dealDamage(targets, user, damage, false, user.element, adventure)} ${dealDamage([user], user, recoilDmg, true, "Untyped", adventure)}`;
```
A template string with only a dynamic value is silly. Instead of the following:
```
`${value}`
```
we should just use `value`.

In the suggestion, I'm using the template string to place the white space joining character. The `+` operator doesn't do that automatically, and is considered a less preferred way of constructing strings because it's less type-safe.

Additionally, I think the recoil damage should be marked as unblockable. Since it's a "cost" to the attack, I don't think asteroids should be able to mitigate their recoil damage via protection or evade.

ntseng commented on source/enemies/asteroid.js:
```diff
@@ -0,0 +1,73 @@
+  priority: 0,
+  effect: (targets, user, isCrit, adventure) => {
+    let damage = user.getPower()+30;
+    let recoilDmg = 20;
```
```suggestion
    const recoilDmg = 20;
```

ntseng commented on source/enemies/asteroid.js:
```diff
@@ -0,0 +1,73 @@
+      resultString += dealDamage([target], user, damage, false, user.element, adventure);
+      changeStagger([target], "elementMatchFoe");
+    }
+    return resultString;
```
```suggestion
    changeStagger(targetList, "elementMatchFoe");
    return `${dealDamage(targetList, user, damage, false, user.element, adventure)} ${getNames([user], adventure)[0]} is downed.`;
```
Both `changeStagger` and `dealDamage` take arrays of targets for this specific use-case.

Also, for the added mention of the asteroid defeating itself, please import the `getNames` function.

ntseng commented on source/enemies/asteroid.js:
```diff
@@ -0,0 +1,73 @@
+    for (let i = 0; i < enemyAttacks; i++) {
+      const liveEnemies = adventure.room.enemies.filter(e => e.hp > 0);
+      targetList.push(liveEnemies[adventure.generateRandomNumber(liveEnemies.length, "battle")]);
+    }
```
```suggestion
    const liveDelvers = adventure.delvers.filter(d => d.hp > 0);
    user.hp = 0; // rule out self from selection
    const liveEnemies = adventure.room.enemies.filter(e => e.hp > 0);
    
    const delverAttacks = Math.floor(numAttacks / user.hp * user.maxHP);    
    const targetList = [];
    for (let i = 0; i < numAttacks; i++) {
      if (i < delverAttacks) {
        targetList.push(liveDelvers[adventure.generateRandomNumber(liveDelvers.length, "battle")]);
      } else {
        targetList.push(liveEnemies[adventure.generateRandomNumber(liveEnemies.length, "battle")]);
      }
    }
```
Changes in order of importance:
1. Moved `liveDelvers` and `liveEnemies` filters out of for loop, as they aren't expected to change between iterations.
2. Made `targetList` a const, as it's never reassigned.
3. Consolidated for loops, removing bug possibilities "calculation of `enemyAttacks` not matching `delverAttacks`" and "a combatant being selected both as a delver and enemy (in case of weird team mechanics in the future)".

ntseng commented on source/enemies/earthlyknight.js:
```diff
@@ -0,0 +1,80 @@
+      const buffIndex = adventure.generateRandomNumber(targetBuffs.length, "battle");
+      const rolledDebuff = targetBuffs[buffIndex];
+      const wasRemoved = removeModifier([target], { name: rolledDebuff, stacks: "all" }).length > 0;
+      if (wasRemoved) {
```
This conditional will always evaluate to `true` because even if `removeModifier` returns an empty array, empty arrays are truthy in JS.

ntseng commented on source/enemies/earthlyknight.js:
```diff
@@ -0,0 +1,80 @@
+    for (target of targets) {
+      resultString += dealDamage([target], user, damage, false, user.element, adventure);
+      const targetBuffs = Object.keys(target.modifiers).filter(modifier => isBuff(modifier));
+      console.log(targetBuffs);
```
Leftover console.log

ntseng commented on source/enemies/earthlyknight.js:
```diff
@@ -0,0 +1,80 @@
+      const rolledDebuff = targetBuffs[buffIndex];
+      const wasRemoved = removeModifier([target], { name: rolledDebuff, stacks: "all" }).length > 0;
+      if (wasRemoved) {
+        resultString += `${rolledDebuff} was lost.`;
```
I think it's important to mention which target is losing their buffs. When the result is going to read XYXYXYX, where X is a damage statement and Y is a debuff loss statement (when some Ys might be missing), it'll be easy to accidentally pair the wrong Xs and Ys.

In that sense, I think the loop should be refactored to a `for(let i = 0; i < targets.length; i++)` so that the index can be used to both iterate over the targets for the debuff removal and find the target's name in the returned array of `getNames(targets, adventure)`.

I think it's an acceptable judgement call to consolidate the damage statements to all be at the beginning via `dealDamage(targets...`, but I'll leave that to you.

ntseng commented on source/enemies/earthlyknight.js:
```diff
@@ -0,0 +1,80 @@
+    let damage = user.getPower() + 75;
+    changeStagger(targets, "elementMatchFoe");
+    let resultString = ""
+    for (target of targets) {
```
```suggestion
    for (const target of targets) {
```
I've you're sticking with a `for(... of ...)` loop (which I recommend against in a further down comment), don't forget to add a declarative to the iterator variable (`target`) so it doesn't get stuck in global scope.

ntseng commented on source/enemies/earthlyknight.js:
```diff
@@ -0,0 +1,80 @@
+  description: `Deal minor ${getEmoji("Earth")} to all foes and stagger them`,
+  priority: 0,
+  effect: (targets, user, isCrit, adventure) => {
+    let damage = user.getPower();
```
Is this intended to deal 0 damage? Non-clone enemies don't usually have base power stats, as those numbers are clearer when baked into the action's damage calculations.

ntseng commented on source/rooms/battle-meteorknight.js:
```diff
@@ -0,0 +1,15 @@
+	generateCombatRoomBuilder([])
+).addEnemy("Earthly Knight", "0.5*n")
+	.addEnemy("Meteor Knight", "1")
+	.addEnemy("Asteroid", "0.25*n-1");
```
I like the use of subtraction in the formula to reduce the impact on low delver count adventures.

ntseng commented on source/enemies/meteorknight.js:
```diff
@@ -0,0 +1,87 @@
+		let damage = user.getPower() + 60;
+		if (isCrit) {
+			damage *= 2;
+		} changeStagger(targets, "elementMatchFoe");
```
```suggestion
		}
		changeStagger(targets, "elementMatchFoe");
```

ntseng commented on source/enemies/meteorknight.js:
```diff
@@ -0,0 +1,87 @@
+}).addAction({
+	name: "Armored Avalanche",
+	element: "Fire",
+	description: "Increases protection of target by up to 50%, then deals large damage, and receives recoil proportional to damage blocked",
```
```suggestion
	description: `deal large ${getEmoji("Fire")} damage that is 33% less effective against protection; receives recoil equal to part of damage blocked`,
```

ntseng commented on source/enemies/meteorknight.js:
```diff
@@ -0,0 +1,87 @@
+		}
+		const recoilDmg = Math.floor(Math.min(totalBlocked, damage) / 2); // half the value of armor applied against damage
+		changeStagger(targets, "elementMatchFoe");
+		return `${protectedCombatants ? joinAsStatement(protectedCombatants.map(c => c.name), false, "is", "were", "braced for the charge!") : ""}${dealDamage(targets, user, damage, false, user.element, adventure)} ${recoilDmg ? user.name + " was repelled!" + dealDamage([user], user, recoilDmg, false, user.element, adventure) : ""}`;
```
Empty array is truthy

ntseng commented on source/enemies/meteorknight.js:
```diff
@@ -0,0 +1,87 @@
+		if (isCrit) {
+			protected = addModifier(adventure.room.enemies.filter(c => c.hp > 0), { name: "protection", stacks: 50 });
+		}
+		if (!(empowered || protected)) {
```
Empty array is truthy

ntseng commented on source/enemies/meteorknight.js:
```diff
@@ -0,0 +1,87 @@
+	description: "Powers up all combatants (friend and foe); Protects non-delvers on crit",
+	priority: 0,
+	effect: (targets, user, isCrit, adventure) => {
+		const empowered = addModifier(adventure.delvers.concat(adventure.room.enemies).filter(c => c.hp > 0), { name: "Power Up", stacks: 50 });
```
50 Power Up seems a bit too fast, that's almost a full basic attack's worth (75)

ntseng reviewed (changes requested):
Great job, this fight looks very fun! Some of the important changes include `adjust for empty array being truthy in JS` and `adding the fight to the labyrinth it was supposed to appear in`.

doubledh committed:
Update source/enemies/asteroid.js

Co-authored-by: Nathaniel Tseng <nathanieltseng@gmail.com>

doubledh committed:
Update source/enemies/meteorknight.js

Co-authored-by: Nathaniel Tseng <nathanieltseng@gmail.com>

doubledh committed:
Update source/enemies/earthlyknight.js

Co-authored-by: Nathaniel Tseng <nathanieltseng@gmail.com>

Commit:
#134: per PR review
- rework armored avalanche
- removed selectNone placeholder combatref
- add selectAll combatReference
- rework bolide burst
- balancing

Commit:
#134: add selectAllOtherAllies/Combatants
- fix earthly knight random

ntseng committed:
convert Action.next to string

ntseng committed:
fix vestigial usages of Combatant.getName

Commit:
Merge branch 'doubledh/issue134' of https://github.com/Imaginary-Horizons-Productions/prophets-of-the-labyrinth into doubledh/issue134
